### PR TITLE
Test how line-height works on <input type=text> and friends

### PR DESCRIPTION
**Pull request by zcorpan**
See https://github.com/whatwg/html/pull/5462

### CODE REVIEW

**Comment by dholbert** [comment]
I think the test expectations are wrong here, as discussed in https://bugzilla.mozilla.org/show_bug.cgi?id=1753537#c3 .

getComputedStyle returns the "resolved value", which is *usually* the computed value but is instead the used value in some cases, as discussed at https://drafts.csswg.org/cssom/#resolved-values .  And `line-height:[not normal]` is one such case where it specifically returns the **used value**.

So: if the test is using `getComputedStyle`, then it should be expecting to get back the **used value** of `line-height` here; and given that the used value is specced as being clamped for these controls, then the test should be expecting that clamping to show up in the returned value.

**Comment by zcorpan** [comment]
@dholbert thanks, you're correct. I'll fix the test.

File: html/rendering/non-replaced-elements/form-controls/input-line-height-computed.html [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+<p><input type=email value=email>
+<p><input type=password value=password></p>
+<script>
+const inputs = document.querySelectorAll('input');
+for (const input of inputs) {
+  test(() => {
+    const usedLineHeight = getComputedStyle(input).lineHeight;
+    assert_not_equals(usedLineHeight, '1px', 'usedLineHeight');
+    assert_not_equals(usedLineHeight, 'normal', 'usedLineHeight');
+  }, `getComputedStyle(<input type=${input.type}>).lineHeight`);
```
Comment (dholbert) [inline html/rendering/non-replaced-elements/form-controls/input-line-height-computed.html:25]: Maybe worth adding a few words of explanation to the test prose here?

e.g. "...should return a used value that corresponds to 'normal' (but should not literally be 'normal')"

File: html/rendering/non-replaced-elements/form-controls/input-line-height-computed.html [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  test(() => {
+    const computedLineHeight = input.computedStyleMap().get('line-height');
+    assert_equals(computedLineHeight.value, 1, 'computedLineHeight.value');
+    assert_equals(computedLineHeight.unit, 'px', 'computedLineHeight.unit');
+  }, `<input type=${input.type}>.computedStyleMap().get('line-height')`);
```
Comment (dholbert) [inline html/rendering/non-replaced-elements/form-controls/input-line-height-computed.html:30]: Same here -- e.g. maybe add "...should not get the same clamping-to-'normal' that the used value gets" or something along those lines?

File: html/rendering/non-replaced-elements/form-controls/input-line-height-computed.html [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+<script src="/resources/testharnessreport.js"></script>
+<style>
+  input { line-height: 1px; }
+</style>
+<p><input type=text value=text>
+<p><input type=tel value=tel>
+<p><input type=search value=search>
+<p><input type=url value=url>
+<p><input type=email value=email>
+<p><input type=password value=password></p>
```
Comment (dholbert) [inline html/rendering/non-replaced-elements/form-controls/input-line-height-computed.html:17]: It makes sense to me that `password` is included here, but the spec seems to imply that it should be excluded right now.

This special clamping behavior is discussed in the first paragraph of https://html.spec.whatwg.org/multipage/rendering.html#the-input-element-as-a-text-entry-widget seems to be talking about just a specific list of `type` values which *does not* include password.

I see you're editing that prose in https://github.com/whatwg/html/commit/c6440d4aedff2f3d167bfe3127f5c5662be6a0ab - maybe it'd be worth extending that edit such that it includes `password` as well, for the clamping bit?  That would make this part of the test valid, I think.  (E.g. *Additionally: for these types as well as for `type="password", [...stuff about clamping...]*)

**Review by dholbert — COMMENTED** [review]
Looks good to me, just a few suggestions (the first of which is actually a suggestion for https://github.com/whatwg/html/commit/c6440d4aedff2f3d167bfe3127f5c5662be6a0ab to make it better-match these tests).

File: html/rendering/non-replaced-elements/form-controls/input-line-height-computed.html (hunk 2 above)
Comment (zcorpan) [reply html/rendering/non-replaced-elements/form-controls/input-line-height-computed.html:30]:
```suggestion
  }, `<input type=${input.type}>.computedStyleMap().get('line-height') should not be affected by the used value clamping`);
```

**Comment by zcorpan** [comment]
@dholbert thanks, clarified the spec in https://github.com/whatwg/html/pull/5462/commits/40c9c5eb122d6682ec4f8ccebcc98918e1e21474

I also changed the test names to describe the expectations.

File: html/rendering/non-replaced-elements/form-controls/input-line-height-computed.html (hunk 2 above)
Comment (dholbert) [reply html/rendering/non-replaced-elements/form-controls/input-line-height-computed.html]: I see this is fixed in https://github.com/web-platform-tests/wpt/pull/32684/commits/bf6cd3966cf4dc1d2015ffae4c260c21e5003f50 - thanks!

File: html/rendering/non-replaced-elements/form-controls/input-line-height-computed.html (hunk 1 above)
Comment (dholbert) [reply html/rendering/non-replaced-elements/form-controls/input-line-height-computed.html]: I see this is fixed in https://github.com/web-platform-tests/wpt/pull/32684/commits/bf6cd3966cf4dc1d2015ffae4c260c21e5003f50 - thanks!

**Review by dholbert — APPROVED** [review]

**Comment by dholbert** [comment]
> @dholbert thanks, clarified the spec in [whatwg/html@40c9c5e](https://github.com/whatwg/html/commit/40c9c5eb122d6682ec4f8ccebcc98918e1e21474)

FWIW I didn't mean to request a spec-clarification along those lines -- I don't think that spec-clarification (saying 'normal' is not expected in the used value) was strictly necessary, since `line-height:normal` is already explicitly defined as setting a numeric used value, here: https://www.w3.org/TR/CSS21/visudet.html#propdef-line-height . So it's by-definition not something that should show up in the used value.

It's fine with the clarification, but if you end up doing further wordsmithing here later on and feel like removing that clarification, I won't object. :)

**Pull request merged**